core/src/main/java/org/infinispan/interceptors/impl/AsyncInvocationStage.java

```diff
@@ -124,39 +117,146 @@ public InvocationStage toInvocationStage(InvocationContext newCtx, VisitableComm
+         if (!currentStage.isDone()) {
+            if (currentStage instanceof BasicAsyncInvocationStage) {
+               // Use the CompletableFuture directly, without creating another AsyncInvocationStage instance
+               currentStage.toCompletableFuture().whenComplete(this);
+               return;
+            } else if (currentStage instanceof AsyncInvocationStage) {
+               AsyncInvocationStage asyncInvocationStage = (AsyncInvocationStage) currentStage;
+               asyncInvocationStage.whenComplete(this);
+               return;
             } else {
-               // if (handler instanceof ExceptionStage)
-               return rv;
+               currentStage = new ExceptionStage(ctx, command, new IllegalStateException(
+                     "Unsupported asynchronous stage type: " + currentStage));
             }
+         }
+
+         // We got a synchronous invocation stage, continue with the next handler
+      }
+   }
+
+   private void completeFromStage(BasicInvocationStage stage) {
+      if (!stage.isDone())
+         throw new IllegalArgumentException("Stage must be done");
+
+      try {
+         Object rv = stage.get();
+         future.complete(rv);
+      } catch (Throwable t) {
+         future.completeExceptionally(t);
+      }
+   }
+
+   private void whenComplete(AsyncInvocationStage otherStage) {
+      if (addHandler(otherStage)) return;
+
+      future.whenComplete(otherStage);
+   }
+
+   /**
+    * The handlers deque: an array-based deque that cannot grow, but it can be frozen so that no more elements
+    * can be added.
+    *
+    * The deque is frozen when someone calls get() or toCompletableFuture(), otherwise we wouldn't know
+    * how many handlers to wait for.
+    * The deque is also frozen when we execute the last handler. Calling compose() later will return a synchronous
+    * invocation stage instead of adding the handler.
+    */
+   // Capacity must be a power of 2
+   private static final int HANDLERS_DEQUE_CAPACITY = 8;
+   @GuardedBy("this")
+   private InvocationComposeHandler[] handlers = new InvocationComposeHandler[HANDLERS_DEQUE_CAPACITY];
+   @GuardedBy("this")
+   private byte mask = HANDLERS_DEQUE_CAPACITY - 1;
+   @GuardedBy("this")
+   private byte head;
+   @GuardedBy("this")
+   private byte tail;
+   @GuardedBy("this")
+   private boolean frozen;
+
+   /**
+    * Stop adding new handlers to this stage
+    */
+   private void lockHandlers() {
+      synchronized (this) {
+         frozen = true;
+      }
+   }
+
+   /**
+    * @return {@code true} if the handler was added, or {@code false} if the handlers deque is frozen
+    */
+   protected boolean addHandler(InvocationComposeHandler composeHandler) {
+      synchronized (this) {
+         if (frozen) {
+            return false;
+         }
+         handlers[tail & mask] = composeHandler;
+         if (((tail + 1 - head) & mask) == 0) {
+            frozen = true;
+            return false;
```

Comment:
Because the add didn't really succeed, as the deque supported only `mask` elements. I've fixed that, and it now supports `HANDLERS_DEQUE_CAPACITY` elements.